### PR TITLE
devices - Ciena RLS 6500

### PR DESCRIPTION
**Pull request by h-barnhart**
Adding the new OS ciena-rls, which adds support for Ciena's 6500 Reconfigurable Line System (RLS) optical networking platform.

This will allow for monitoring of optical line system transmit and receive, optical return loss, amplifier status and state, and temperature monitoring.

YAML discovery was used as much as possible. I had to create PHP for the sensors because Ciena used two byte strings at the end of the OID as the index and as far I could tell the YAML would only accept one byte string as an index.

Adds the appropriate MIB files which also updates CIENA-SMI.

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

RLS

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by h-barnhart** [comment]
Running into an issue with the snmpsim data causing tests to fail in the port discovery. The tests expect ifVLAN to be empty ('') but its getting a null. I'm not sure how to resolve that.

**Comment by PipoCanaja** [comment]
Hi @h-barnhart 
Looking at the mib : 
```
rlsInventoryAmpsEntry OBJECT-TYPE
    SYNTAX      RlsInventoryAmpsEntry
    MAX-ACCESS  not-accessible
    STATUS      current
    DESCRIPTION 
        ""
    INDEX       { rlsInventoryAmpsSlotName, rlsInventoryAmpsName}
    ::= { rlsInventoryAmpsTable 1 }
```
For instance : 
"sensor_oid": ".1.3.6.1.4.1.1271.4.1.1.1.2.1.1.1.48.1.49.7.66.111.111.115.116.101.114",
"sensor_index": "rlsInventoryAmpsInCurrPower.1.49.7.66.111.111.115.116.101.114",
"sensor_descr": "Slot 1 Booster",

That can be decoded as "1 char ASC(49) which is "1" and 7 chars ASC( 66.111.111.115.116.101.114 ) which is "Booster". 

So it is 1 byte string, but dual indexed string ... And you are right, current code only support the 1st index string (in that case "1".). Remaining data is ignored. 

I lack time these days, so unless a miracle happens, your PHP code will make it. If by chance I can improve the index parsing, you may be able to YAML all :)

**Comment by h-barnhart** [comment]
> So it is 1 byte string, but dual indexed string ... And you are right, current code only support the 1st index string (in that case "1".). Remaining data is ignored.
> 
> I lack time these days, so unless a miracle happens, your PHP code will make it. If by chance I can improve the index parsing, you may be able to YAML all :)

Right on. The flip slide is MIB does have entries in the table for both parts of the index that make up the byte string (slot ID and slot name), but doesn't allow access to those entries. If they did I probably wouldn't have had to code as much.

**Comment by h-barnhart** [comment]
>1) LibreNMS\Tests\YamlSchemaTest::testDiscoveryDefinitionSchema
>ciena-rls.yaml does not validate. Violations:
>[modules.sensors.temperature.data[0].num_oid] Does not match the regex pattern ^(\.\d+)+\.?\{\{ \$index \}\}(\.\d+)*$

Are you not allowed to use index_string anymore?

**Comment by h-barnhart** [comment]
> `1) LibreNMS\Tests\YamlSchemaTest::testDiscoveryDefinitionSchema ciena-rls.yaml does not validate. Violations: [modules.sensors.temperature.data[0].num_oid] Does not match the regex pattern ^(\.\d+)+\.?\{\{ \$index \}\}(\.\d+)*$ `
> 
> Are you not allowed to use index_string anymore in YAML discovery? The index value is a string which identifies the slot on the device where the temperature sensor is located. I'm using that in the description of the temp sensor, as there may be multiple versions of the same hardware component in different slots.
> 
> If I can't use $index_string, is there a way in the YAML to get the string value of the index? If not I'll code it out in PHP.

I went ahead made a php script to get past the tests. All green now.

**Comment by PipoCanaja** [comment]
> `1) LibreNMS\Tests\YamlSchemaTest::testDiscoveryDefinitionSchema ciena-rls.yaml does not validate. Violations: [modules.sensors.temperature.data[0].num_oid] Does not match the regex pattern ^(\.\d+)+\.?\{\{ \$index \}\}(\.\d+)*$ `
> 
> Are you not allowed to use index_string anymore in YAML discovery? The index value is a string which identifies the slot on the device where the temperature sensor is located. I'm using that in the description of the temp sensor, as there may be multiple versions of the same hardware component in different slots.
> 
> If I can't use $index_string, is there a way in the YAML to get the string value of the index? If not I'll code it out in PHP.

Should be, it is still used by some YAML files. But in your case, index_string was not the solution. 
Index string converts an index to numeric in case it was a string. You have numerical index and would like to convert it to string, to use it in description of the sensor. So basically, you don't need index_string (but thanx for pointing out that there is something to fix here).

**Comment by PipoCanaja** [comment]
Fixed `index_string`for single string indexed OIDs. Meaning you can restore your YAML for temperatures for instance.

**Comment by PipoCanaja** [comment]
Now, index_string can correctly handle an index like "678.hw87" and convert it to the OID representation of it (with length for each of the 2 subindexes and conversion to DEC).

File: includes/definitions/ciena-rls.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+os: ciena-rls
+text: 'Ciena Reconfigurable Line System'
+icon: ciena
+over:
+    - { graph: device_bits, text: Traffic }
+type: network
+mib_dir: ciena
+discovery_modules:
+    bgp-peers: false
```
Comment (PipoCanaja) [inline includes/definitions/ciena-rls.yaml:9]: Is there a reason to disable these ? The rule is to keep everything enabled unless it breaks something.

File: includes/definitions/ciena-rls.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    cisco-vrf-lite: false
+    fdb-table: false
+    hr-device: false
+    mpls: false
+    ports-stack: false
+    stp: false
+    ucd-diskio: false
+    vlans: false
+poller_modules:
+    bgp-peers: false
```
Comment (PipoCanaja) [inline includes/definitions/ciena-rls.yaml:19]: same here

**Review by PipoCanaja — CHANGES_REQUESTED** [review]

**Comment by PipoCanaja** [comment]
Seems that issue https://community.librenms.org/t/juniper-loss-not-polling-correctly/16628/3 is now solved

File: LibreNMS/Device/YamlDiscovery.php [hunk 3]
```diff
@@ -188,6 +188,11 @@ public static function replaceValues($name, $index, $count, $def, $pre_cache)
             $variables = [
                 'index' => $index,
                 'count' => $count,
+                // we compute a numOid compatible version of index
+                // string length followed by ASCII of each char.
+                'index_string' => implode('.', array_map(function ($index) {
```
Comment (PipoCanaja) [inline LibreNMS/Device/YamlDiscovery.php:193]: This correct implementation is the fix for this old issue:  https://community.librenms.org/t/juniper-loss-not-polling-correctly/16628/3

**Comment by PipoCanaja** [comment]
Should be possible to do the states in YAML as well, but tomorrow will be another day.

**Comment by PipoCanaja** [comment]
> > So it is 1 byte string, but dual indexed string ... And you are right, current code only support the 1st index string (in that case "1".). Remaining data is ignored.
> > I lack time these days, so unless a miracle happens, your PHP code will make it. If by chance I can improve the index parsing, you may be able to YAML all :)
> 
> Right on. The flip slide is MIB does have entries in the table for both parts of the index that make up the byte string (slot ID and slot name), but doesn't allow access to those entries. If they did I probably wouldn't have had to code as much.

Currently, with the new changes in the code, the only issue is that description is `Slot 1.Booster` instead of `Slot 1 Booster`. I have no way to remove the dot in between. But if you can live with that, YAML is a looooooot more compact and easier to maintain.

**Comment by PipoCanaja** [comment]
All done @h-barnhart 
Let me know for the discovery and poller modules, and if we can clean them all, it will be ready to merge.

**Comment by h-barnhart** [comment]
Those poller and discovery modules aren't used, I was trying to be thorough. I put them back and push a new change.

Thanks for looking into the YAML. If you want me to put the temp sensor back in YAML I can and try out the DBM and State sensors as well. I was hoping to get this in before the next full release though.

**Comment by h-barnhart** [comment]
> Those poller and discovery modules aren't used, I was trying to be thorough. I put them back and push a new change.
> 
> Thanks for looking into the YAML. If you want me to put the temp sensor back in YAML I can and try out the DBM and State sensors as well. I was hoping to get this in before the next full release though.

Oh, you did all that. Thanks

I re-enabled the poller and discovery modules as requested.

**Review by PipoCanaja — APPROVED** [review]
LGTM, thanx for the PR.

**Pull request merged**